Assignment_4.py

```diff
@@ -0,0 +1,318 @@
+                name=int(curr_group),
+                box_visible=True,
+                meanline_visible=True,
+            )
+        )
+    fig_1.update_layout(
+        title="Continuous Response by Categorical Predictor",
+        xaxis_title="Groupings",
+        yaxis_title="Response",
+    )
+    # fig_1.show()
+    fig_1.write_html(
+        file=filename,
+        include_plotlyjs="cdn",
+    )
+
+
+def generate_dist_plot(df, col, filename):
+    group_labels = ["0", "1"]
+    # Create distribution plot with custom bin_size
+    M = df[df["target"] == 0][col]
+    B = df[df["target"] == 1][col]
+    fig_1 = ff.create_distplot([M, B], group_labels, bin_size=0.2)
+    fig_1.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title="Predictor",
+        yaxis_title="Distribution",
+    )
+    # fig_1.show()
+    fig_1.write_html(
+        file=filename,
+        include_plotlyjs="cdn",
+    )
+
+
+def generate_scatter_plot(df, col, filename):
+    fig = px.scatter(x=df[col], y=df["target"], trendline="ols")
+    fig.update_layout(
+        title="Continuous Response by Continuous Predictor ",
+        xaxis_title="Predictor",
+        yaxis_title="Response",
+    )
+    # fig.show()
+    fig.write_html(
+        file=filename,
+        include_plotlyjs="cdn",
+    )
+
+
+max_bin = 10
+
+
+def calculate_mean_of_response(df, col, pop_prop_1, n=max_bin):
+    print(col)
+    d1 = pd.DataFrame({"X": df[col], "Y": df["target"], "Bucket": pd.qcut(df[col], 10)})
+    d2 = d1.groupby("Bucket", as_index=True)
+    d3 = pd.DataFrame({}, index=[])
+    d3["LowerBin"] = d2.min().X
+    d3["UpperBin"] = d2.max().X
+    d3["COUNT"] = d2.count().Y
+    pop_prop = d3["COUNT"] / len(df)
+    d3["BinMean"] = d2.sum().Y / d3.COUNT
+    d3["Mean_sq_diff"] = (d3.BinMean - pop_prop_1) ** 2
+    d3["Mean_sq_diffW"] = d3.Mean_sq_diff * pop_prop
+    return d3["Mean_sq_diff"].sum(), d3["Mean_sq_diffW"].sum()
+
+
+def calculate_mean_of_response_cat(df, col, pop_prop_1):
+    # bins = len(np.unique(df[col].values))
+    d1 = pd.DataFrame({"X": df[col], "Y": df["target"]}).groupby(df[col])
+    d3 = pd.DataFrame({}, index=[])
+    d3["COUNT"] = d1.count().Y
+    pop_prop = d3["COUNT"] / len(df)
+    d3["BinMean"] = d1.sum().Y / d3.COUNT
+    d3["Mean_Sq_Diff"] = (d3.BinMean - pop_prop_1) ** 2
+    d3["Mean_Sq_DiffW"] = d3.Mean_Sq_Diff * pop_prop
+    return d3["Mean_Sq_Diff"].sum(), d3["Mean_Sq_DiffW"].sum()
+
+    print("calculate_mean_of_response_cat")
+
+
+def read_breast_cancer_data():
+    inp_data = pd.read_csv("./Data_Files/wisconsin-breast-cancer-dataset.csv")
+    inp_data.drop(["Unnamed: 32", "id"], axis=1, inplace=True)
+    return inp_data
+
+
+def main(fname, r_name):
+    if fname == "":
+        inp_data = read_breast_cancer_data()
+    else:
+        try:
+            inp_data = pd.read_csv(fname)
+        except FileNotFoundError:
+            inp_data = read_breast_cancer_data()
```

Comment:
I really like the use of try and except block here, this way a user can either test it on the data he/she has. Or if someone is running it for the first time, they can run it on an already defined dataset.